template: support templating of task runtime parameters

stevvooe opened the pull request:
To better support common use cases that require injected service and
task information, we now treat ContainerSpec fields as templates. A
common, simplified context is defined that contains commonly needed
values, with access to names, identifiers and labels.

This initial PR focuses on supporting mounts and environment variables.
By supporting these, users will be able to locate mounts based on task
context or service name, unlocking many use cases that were previously
challenging or impossible. Environment variables open up further
possibilities.

Signed-off-by: Stephen J Day stephen.day@docker.com

cc @aluzzardi @mgoelzer @ehazlett @NathanMcCauley @diogomonica

aaronlehmann commented:
This is pretty cool.

aaronlehmann commented:
There are some golint issues around copying a structure with a lock inside it.

I don't have a strong opinion about the template cache. I would think that templates are relatively cheap to parse (sort of like a printf format string), and if we didn't cache them, we would only need to reparse when handling a new task. This is not exactly something we'd be doing in an inner loop, so the caching probably doesn't buy us much. But maybe template parsing has more overhead than I naively expect.

aluzzardi commented:
> I don't have a strong opinion about the template cache

Yeah, template compilation is dirt cheap compared to prepping a container

stevvooe commented:
I'll remove the template cache. I was breaking my own rules about premature optimization and felt dirty while doing it. If we prove a bottleneck here, we'll take care of it later. The actual answer might be to "de-feature" the template system, since complex templating isn't really needed.

diogomonica commented:
Looks pretty cool. Will this bear any relation with config management?

stevvooe commented:
@diogomonica This is the engine for configuration management:

- Imagine a new mount type that carries a file that can be evaluated as a template.
- Imagine having a secret template function to access to secrets within a template
- Imagine wiring this context to the evaluated resources to have the template re-expanded on resource change
- Imagine hooking in new sources for populating template variables

tlvenn commented:
That is awesome @stevvooe !

My usecase would be to create a dir mounted as tmpfs and create a file in it containing credentials pulled from Vault that my service can access transparently. 

Can't wait for this to land.

diogomonica commented:
@tlvenn for this release all the secrets will be stored on the swarm managers, and will have to be added/removed/managed directly from the `docker secret` command. Later on we would love to support adding externally managed secrets that are imported from external secret stores and distributed throughout all the nodes.

dnephin commented:
This looks great.

What do you think about supporting `ContainerSpec.Name` and `ContainerSpec.Hostname` ?

stevvooe commented:
@dnephin There is no `ContainerSpec.Name` field. We'd have to integrate `Annotations` and work to fix the backwards compatibility issues.

`ContainerSpec.Hostname` is not yet merged, but I think we can go ahead and merge it and add it without issue. The main problem with it, as I found out the other day, is that `Hostname` isn't just the hostname. It is parsed into both `Hostname` and `DomainName`, so we have to honor that model.

thaJeztah commented:
Based on all the "ooooh" and "ahhhh" above, shall I move this to "code review"?

![](https://s-media-cache-ak0.pinimg.com/236x/b5/5c/51/b55c5176b934165f58a07073614d0c8c.jpg)

aluzzardi commented:
LGTM

Please rebase

stevvooe commented:
Added support for parameterizing `hostname`.

tlvenn commented:
@diogomonica Can you link the Issue/PR/Proposal if any that introduce `docker secret` command ? I would love to learn more about it. Thanks !

stevvooe commented:
A few more changes:
- removed support for `Task.Labels`. These aren't really of interest to the user. They can set container labels manually or template with task labels. We may allow templating of container labels in the future and this limitation preserves that choice.
- Added some pedantic units tests ensuring that labels from the task.annotations don't pollute the templates.

thaJeztah commented:
ping @aaronlehmann @diogomonica PTAL, let's see if we can get this into 1.13 :heart:

aaronlehmann commented on template/expand.go:
```diff
@@ -0,0 +1,115 @@
+func ExpandContainerSpec(t *api.Task) (*api.ContainerSpec, error) {
+	container := t.Spec.GetContainer()
+	if container == nil {
+		return nil, nil
```
Should this return an error?

aaronlehmann commented on template/expand.go:
```diff
@@ -0,0 +1,115 @@
+		}
+
+		n[k] = v
+
```
nit: blank line looks out of place

stevvooe commented on template/expand.go:
```diff
@@ -0,0 +1,115 @@
+				return values, err
+			}
+
+			entry = fmt.Sprintf("%s=%s", entry, expanded)
```
Doesn't that require two intermediates?

stevvooe commented on template/context.go:
```diff
@@ -0,0 +1,65 @@
+	ctx.Node.ID = t.NodeID
+
+	ctx.Task.ID = t.ID
+	ctx.Task.Name = fmt.Sprintf("%s.%d.%s", t.ServiceAnnotations.Name, t.Slot, t.ID)
```
Do as I say, not as I do? :trollface: 🐹 

We still need to centralize the naming. Is this okay for now or do want me to introduce the `naming` package?

stevvooe replied:
@aluzzardi Seems outside the scope of this PR. I've asked for this in 6 other PRs and I'm not gating this feature on that addition if no one else is willing to.

stevvooe replied:
Naming package: https://github.com/docker/swarmkit/pull/1694.

aluzzardi commented:
LGTM

Needs rebase
